Clipboard: Removed phantomjs dependency

ossdev07 opened the pull request:
Phantomjs is an unmaintained project, removed phantomjs dependency from package.json and karma.conf.js.Added chrome headless support in package.json and karma.conf.js.
The dist/clipboard.js and dist/clipboard.min.js file got updated whenever we install and test the package.

Signed-off-by: ossdev <ossdev@puresoftware.com>

ossdev07 commented:
Hi,

Added custom launcher support for headless chrome.
-> Reason for using --no-sandbox flag:

For aarch64 chrome binary is not present, we are using chromium binary to launch the Chrome.
Security is one of the most important goals for Chromium, and --no-sandbox flag provides that.
When trying to use chromium as root user like in docker container then only the requirement of --no-sandbox flag is coming
Without the flag, chromium gives error as root user-
~~~
[28769:28769:0925/163702.505790:ERROR:zygote_host_impl_linux.cc(89)] Running as root without --no-sandbox is not supported. See https://crbug.com/638180.
~~~
Thus, we need to give the --no-sandbox flag when we launch chrome with chromium binary.

zenorocha commented:
Hey @ossdev07! I love that idea, thanks for sending this PR.

Looks like Travis is complaining about it, could you make it work on Travis before I merge? Thanks!

Ruffio commented:
@ossdev07 Will you update so it can be merged?

Ruffio commented:
@ossdev07 what do you sat?

zenorocha commented:
Thanks!

ossdev07 commented:
Hi @zenorocha 

I have been working on the failing 3 test cases of the package and would really appreciate if you may guide me through some of the aspects of the test cases.

```
    describe('#static isSupported', () => {
        it('should return the support of the given action', () => {
            assert.equal(Clipboard.isSupported('copy'), false);
            assert.equal(Clipboard.isSupported('cut'), false);
        });

        it('should return the support of the cut and copy actions', () => {
            assert.equal(Clipboard.isSupported(), false);
        });
    });
```
The above test case is causing the 2 assertion issue in the package, in your view can we modify the test case or skip it if it is not essential for the package.

DanielRuf commented:
`window.Clipboard` is a native part of current browsers so this is expected to fail imo.

Should be renamed to `ClipboardJS`.

Ruffio commented:
@ossdev07 and @zenorpcha what is status? Are we coming to an end?

DanielRuf commented:
> ### This branch has conflicts that must be resolved
> Only those with [write access](https://help.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.
> #### Conflicting files
> `dist/clipboard.js`
> `dist/clipboard.min.js`
> `package.json`

ossdev07 commented:
Hi 

I have updated the src/clipboard.js and src/clipboard-action.js files to resolve the issue of failing 3 test cases with headless chrome, apparently, the phantomjs and chrome work differently that is why I have updated the file.
Also, for the clear selection function, it was not taking the focus off selected element so I have added to blur the active element.
The dist/clipboard.js and dist/clipboard.min.js get updated itself whenever we run install and test the package if you want I can remove these files from the pull request for better clarity.

DanielRuf commented:
That does not really answer my question.

ossdev07 commented:
@DanielRuf 

The changes which I did regarding the check for false inside src/clipboard.js is because of the following reason:

- Actually in phantomjs document.queryCommandSupported gives output as function where as !!document.queryCommandSupported gives as false.

~~~
25 03 2019 14:19:25.046:INFO [PhantomJS 2.1.1 (Linux 0.0.0)]: Connected on socket tn_0845WgY2jdzdqAAAA with id 49411648
LOG: 'document.queryCommandSupported = function queryCommandSupported() {
    [native code]
}'
LOG: '!document.queryCommandSupported = false'
LOG: '!!document.queryCommandSupported = true'
LOG: '!!document.queryCommandSupported(copy) = false'
LOG: 'document.queryCommandSupported = function queryCommandSupported() {
    [native code]
}'
LOG: '!document.queryCommandSupported = false'
LOG: '!!document.queryCommandSupported = true'
LOG: '!!document.queryCommandSupported(cut) = false'
LOG: 'document.queryCommandSupported = function queryCommandSupported() {
    [native code]
}'
LOG: '!document.queryCommandSupported = false'
LOG: '!!document.queryCommandSupported = true'
LOG: '!!document.queryCommandSupported(copy) = false'
LOG: '!!document.queryCommandSupported(cut) = false'
PhantomJS 2.1.1 (Linux 0.0.0): Executed 26 of 26 SUCCESS (0.052 secs / 0.018 secs)
TOTAL: 26 SUCCESS
~~~

- But in chrome document.queryCommandSupported gives output as function but !! document.queryCommandSupported gives as true.

~~~
25 03 2019 14:44:07.672:INFO [HeadlessChrome 73.0.3683 (Ubuntu 0.0.0)]: Connected on socket KVb6qeAIzV4c6AaOAAAA with id 19968799
LOG: 'document.queryCommandSupported = function queryCommandSupported() { [native code] }'
LOG: '!document.queryCommandSupported = false'
LOG: '!!document.queryCommandSupported = true'
LOG: '!!document.queryCommandSupported(copy) = true'
LOG: 'document.queryCommandSupported = function queryCommandSupported() { [native code] }'
LOG: '!document.queryCommandSupported = false'
LOG: '!!document.queryCommandSupported = true'
LOG: '!!document.queryCommandSupported(cut) = true'
LOG: 'document.queryCommandSupported = function queryCommandSupported() { [native code] }'
LOG: '!document.queryCommandSupported = false'
LOG: '!!document.queryCommandSupported = true'
LOG: '!!document.queryCommandSupported(copy) = true'
LOG: '!!document.queryCommandSupported(cut) = true'
HeadlessChrome 73.0.3683 (Ubuntu 0.0.0): Executed 26 of 26 SUCCESS (0.039 secs / 0.022 secs)
TOTAL: 26 SUCCESS
~~~

- Regarding the use of TinyEmitter, I have not additionally added this in the file as I have already mentioned this file gets updated by itself whenever we try to test the package, hence this is being introduced by itself in the file and I have not added it from my end.

- I can remove the dist/clipboard.js and dist/clipboard.min.js files from the pull request as the changes in those files are not suggested from my end and are introduced itself by testing and running the package.

Please revert if something is still not clear I will happy to explain this again.

DanielRuf commented:
> * Actually in phantomjs document.queryCommandSupported gives output as function where as !!document.queryCommandSupported gives as false.

See https://developer.mozilla.org/en-US/docs/Web/API/Document/queryCommandSupported
> document.queryCommandSupported gives as true.

Please do not commit the `dist` files. Remove them in another commit and squash both commits.

> #### Conflicting files
> `dist/clipboard.js`
> `dist/clipboard.min.js`
> `package.json`

ossdev07 commented:
Hi @zenorocha @Ruffio 

The Travis-ci checks are now passing could you please have a look at it, and check it at your end. Please do revert if something is not clear I will be happy to help.
Waiting for the phantomjs dependency to be removed from the package.
Thanks again for your support and patience.

DanielRuf commented:
In the meantime you can use https://github.com/DanielRuf/phantomjs-mocked

rhenwood-arm commented:
I would like to see the PhantomJS dependency removed. Not only is it unmaintained, but PhantomJS is also not available on AArch64 - so the test suite does not pass on my AArch64 platform.

ossdev07 commented:
Ping @Ruffio @zenorocha 

Please, could you accept this PR?

ossdev07 commented:
Ping @zenorocha @Ruffio could you please have a look at this PR?

ossdev07 commented:
@zenorocha @Ruffio @DanielRuf Can someone have a look at this PR, I will be happy to update the patch if you have any suggestions.

DanielRuf commented:
I'm not a collaborator or member so I can not do any review or merge this.

Ruffio commented:
I'm not a collaborator or member so I can not do any review or merge this.

Ruffio commented:
@zenorocha can this PR be accepted? Else write to @ossdev07 what should be changed for it to get merged ;-)

ogonkov commented on src/clipboard.js:
```diff
@@ -86,10 +86,10 @@ class Clipboard extends Emitter {
     static isSupported(action = ['copy', 'cut']) {
         const actions = (typeof action === 'string') ? [action] : action;
-        let support = !!document.queryCommandSupported;
+        let support = !document.queryCommandSupported;
```
There likely should be

```js
'queryCommandSupported' in document && typeof document.queryCommandSupported === 'function'
```

DanielRuf replied:
Why make it more complicated again? I would keep `!!` to convert it to a boolean.

ogonkov replied:
It's not that complicated, but make check more obvious

DanielRuf replied:
I don't think that this should be part of this refactor / change PR but another.

ogonkov replied:
You right

zenorocha commented:
Thanks @ossdev07 and everyone else in this thread for your patience.
